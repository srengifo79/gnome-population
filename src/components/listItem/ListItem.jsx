import { useState } from "react";
import styled from "styled-components";
import Collapse from "@mui/material/Collapse";
import { CircularProgress } from "@mui/material";

import { colors } from "../../theme/colors";
import breakpoints from "../../theme/breakpoints";

const StyledListItem = styled.div`
  display: flex;
  flex-direction: row;
  min-height: 8rem;
  padding: 0.3125rem;
  background-color: ${colors.darkBlue};
  color: ${colors.grayWhite};

  .avatar {
    height: 100%;
    flex: 1;
    margin-right: 1em;
    text-align: center;

    ${breakpoints.xlarge} {
      max-width: 19rem;
    }

    ${breakpoints.xxlarge} {
      max-width: 25rem;
    }

    img {
      max-height: 100%;
      width: 100%;
    }
  }

  .info-container {
    flex: 2;

    span {
      display: block;
    }

    .header {
      margin-bottom: 1em;

      h3 {
        color: ${colors.white};
        margin: unset;
      }

      span {
        font-size: 0.875em;
      }
    }

    .general-data {
      span {
        margin-bottom: 0.3em;
      }
    }
  }
`;

const StyledDropdowBox = styled.div`
  margin: 0 0.5rem;
  border-radius: 0.3em;
  color: ${colors.white};
  background: ${colors.blueGrayGradient};

  .MuiCollapse-wrapper {
    padding: 0.5em;
  }

  .hair-color {
    h4 {
      display: inline;
      margin: unset;
      margin-right: 1em;
    }

    span {
      color: ${colors.grayWhite};
    }
  }

  .list {
    span, li {
      color: ${colors.grayWhite};
    }
  }

  }
`;

const DropdowBox = ({ isExpanded, hair_color, professions, friends }) => (
  <StyledDropdowBox isExpanded={isExpanded}>
    <Collapse in={isExpanded}>
      <>
        <div className="hair-color card">
          <h4>Hair Color:</h4>
          <span>{hair_color}</span>
        </div>
        <div className="list">
          <h4>Professions: </h4>
          {professions.length > 0 ? (
            <ul>
              {professions.map((prof) => (
                <li key={prof}>{prof}</li>
              ))}
            </ul>
          ) : (
            <span>No known professions</span>
          )}
        </div>
        <div className="list">
          <h4>Friends: </h4>
          {friends.length > 0 ? (
            <ul>
              {friends.map((friend) => (
                <li key={friend}>{friend}</li>
              ))}
            </ul>
          ) : (
            <span>No known friends</span>
          )}
        </div>
      </>
    </Collapse>
  </StyledDropdowBox>
);

const Header = ({ id, age, name, height, weight, thumbnail, onClick }) => {
  const [isLoadingImg, setIsLoadingImg] = useState(true);
  return (
    <StyledListItem onClick={onClick} data-testid="listItemContainer">
      {thumbnail && (
        <div className="avatar">
          {isLoadingImg && <CircularProgress />}
          <img
            src={thumbnail}
            alt="gnome avatar"
            onLoad={() => setIsLoadingImg(false)}
          />
        </div>
      )}
      <div className="info-container">
        <div className="header">
          <h3>{name}</h3>
          <span>ID: {id}</span>
        </div>
        <div className="general-data">
          <span>Age: {age}</span>
          <span>Height: {height}</span>
          <span>Weight: {weight}</span>
        </div>
      </div>
    </StyledListItem>
  );
};

const ListItem = (props) => {
  const [isExpanded, setIsExpanded] = useState(false);

  const handleToggle = () => {
    setIsExpanded(!isExpanded);
  };

  return (
    <>
      <Header {...props} onClick={handleToggle} />
      {props.collapsable && <DropdowBox {...props} isExpanded={isExpanded} />}
    </>
  );
};

export default ListItem;
