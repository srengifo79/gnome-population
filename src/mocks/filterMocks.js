export const filterMock = {
  professions: [
    " Tinker",
    "Baker",
    "Blacksmith",
    "Brewer",
    "Butcher",
    "Carpenter",
    "Cook",
    "Farmer",
    "Gemcutter",
    "Leatherworker",
    "Marble Carver",
    "Mason",
    "Mechanic",
    "Medic",
    "Metalworker",
    "Miner",
    "Potter",
    "Prospector",
    "Sculptor",
    "Smelter",
    "Stonecarver",
    "Tailor",
    "Tax inspector",
    "Woodcarver",
  ],
  hairColors: ["Black", "Gray", "Green", "Pink", "Red"],
  ageRange: [30, 379],
  heightRange: [91.41726, 129.37299],
  weightRange: [35.00214, 44.997665],
};
