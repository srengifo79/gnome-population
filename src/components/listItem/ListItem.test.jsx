import React from "react";
import { screen, render, cleanup, fireEvent } from "@testing-library/react";
import { toMatchDiffSnapshot } from "snapshot-diff";

import ListItem from "./ListItem";
import { gnomesDataResponseMock } from "../../mocks/apiMocks";

expect.extend({ toMatchDiffSnapshot });

afterEach(cleanup);

it("Initial snapshop", () => {
  const { asFragment } = render(
    <ListItem {...gnomesDataResponseMock.Brastlewark[0]} />
  );
  expect(asFragment()).toMatchSnapshot();
});

it("Expanded snapshot", () => {
  const { asFragment } = render(
    <ListItem {...gnomesDataResponseMock.Brastlewark[0]} collapsable={true} />
  );
  const initalRender = asFragment();
  const listItemElement = screen.getByTestId("listItemContainer");

  fireEvent.click(listItemElement);

  expect(initalRender).toMatchDiffSnapshot(asFragment());
});
