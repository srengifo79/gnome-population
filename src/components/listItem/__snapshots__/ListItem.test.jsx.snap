// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expanded snapshot 1`] = `
"Snapshot Diff:
- First value
+ Second value

@@ -60,12 +60,12 @@
    </div>
    <div
      class=\\"sc-gsDKAQ iarUdB\\"
    >
      <div
-       class=\\"MuiCollapse-root MuiCollapse-vertical MuiCollapse-hidden css-bz4dnt-MuiCollapse-root\\"
-       style=\\"min-height: 0px;\\"
+       class=\\"MuiCollapse-root MuiCollapse-vertical css-1fkrdiv-MuiCollapse-root\\"
+       style=\\"min-height: 0px; height: 0px; transition-duration: 300ms;\\"
      >
        <div
          class=\\"MuiCollapse-wrapper MuiCollapse-vertical css-smkl36-MuiCollapse-wrapper\\"
        >
          <div"
`;

exports[`Initial snapshop 1`] = `
<DocumentFragment>
  <div
    class="sc-bdvvtL bPnxdX"
    data-testid="listItemContainer"
  >
    <div
      class="avatar"
    >
      <span
        class="MuiCircularProgress-root MuiCircularProgress-indeterminate MuiCircularProgress-colorPrimary css-18lrjg1-MuiCircularProgress-root"
        role="progressbar"
        style="width: 40px; height: 40px;"
      >
        <svg
          class="MuiCircularProgress-svg css-1idz92c-MuiCircularProgress-svg"
          viewBox="22 22 44 44"
        >
          <circle
            class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate css-176wh8e-MuiCircularProgress-circle"
            cx="44"
            cy="44"
            fill="none"
            r="20.2"
            stroke-width="3.6"
          />
        </svg>
      </span>
      <img
        alt="gnome avatar"
        src="http://www.publicdomainpictures.net/pictures/10000/nahled/thinking-monkey-11282237747K8xB.jpg"
      />
    </div>
    <div
      class="info-container"
    >
      <div
        class="header"
      >
        <h3>
          Tobus Quickwhistle
        </h3>
        <span>
          ID: 0
        </span>
      </div>
      <div
        class="general-data"
      >
        <span>
          Age: 306
        </span>
        <span>
          Height: 107.75835
        </span>
        <span>
          Weight: 39.065952
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
